modules/videoio/src/cap_gstreamer.cpp

```diff
@@ -388,23 +495,142 @@ bool GStreamerCapture::grabFrame()
+    {
+        // bail out in no caps
+        CV_LOG_ERROR(NULL, "GStreamer: GST_CAPS_IS_SIMPLE(frame_caps) check is failed");
+        return false;
+    }
+
+    GstAudioInfo info = {};
+    gboolean audio_info_res = gst_audio_info_from_caps(&info, frame_caps);
+    if (!audio_info_res)
+    {
+        CV_Error(Error::StsError, "GStreamer: gst_audio_info_from_caps() is failed. Can't handle unknown layout");
+    }
+    int bpf = GST_AUDIO_INFO_BPF(&info);
+
+    GstStructure* structure = gst_caps_get_structure(frame_caps, 0);  // no lifetime transfer
+    if (!structure)
+    {
+        CV_LOG_ERROR(NULL, "GStreamer: Can't query 'structure'-0 from GStreamer sample");
+        return false;
+    }
+
+    const gchar* name_ = gst_structure_get_name(structure);
+    if (!name_)
+    {
+        CV_LOG_ERROR(NULL, "GStreamer: Can't query 'name' from GStreamer sample");
+        return false;
+    }
+    std::string name = toLowerCase(std::string(name_));
+
+    GstBuffer* buf = gst_sample_get_buffer(sample);
+    if (!buf)
+        return false;
+    GstMapInfo map_info = {};
+    if (!gst_buffer_map(buf, &map_info, GST_MAP_READ))
     {
+        CV_LOG_ERROR(NULL, "GStreamer: Failed to map GStreamer buffer to system memory");
         return false;
     }
+    ScopeGuardGstMapInfo map_guard(buf, &map_info);
+    if (name == "audio/x-raw")
+    {
+        const gchar* format_ = gst_structure_get_string(structure, "format");
+        if (!format_)
+        {
+            CV_LOG_ERROR(NULL, "GStreamer: Can't query 'format' of 'video/x-raw'");
+            return false;
+        }
+        std::string format = toUpperCase(std::string(format_));
+        cv::Mat data;
+        if (format == "S8")
+        {
+            Mat(map_info.size/bpf, nAudioChannels, CV_8S, map_info.data).copyTo(audioFrame);
+            return true;
+        }
+        if (format == "S16LE")
+        {
+            Mat(map_info.size/bpf, nAudioChannels, CV_16S, map_info.data).copyTo(audioFrame);
+            return true;
+        }
+        if (format == "S32LE")
+        {
+            Mat(map_info.size/bpf, nAudioChannels, CV_32S, map_info.data).copyTo(audioFrame);
+            return true;
+        }
+        if (format == "F32LE")
+        {
+            Mat(map_info.size/bpf, nAudioChannels, CV_32F, map_info.data).copyTo(audioFrame);
+            return true;
+        }
+    }
+
+    CV_Error_(Error::StsNotImplemented, ("Unsupported GStreamer layer type: %s", name.c_str()));
+}
+
+bool GStreamerCapture::retrieveAudioFrame(int index, OutputArray dst)
+{
+    if (index > nAudioChannels + audioBaseIndex)
+        return false;
+
+    cv::Mat data;
+    switch (outputAudioFormat)
+    {
+        case CV_8S:
+            data = cv::Mat(1, audioFrame.rows, CV_8S);
+            for (int i = 0; i < audioFrame.rows; i++)
+                data.at<char>(0,i) = audioFrame.at<char>(i, index-audioBaseIndex);
+            data.copyTo(dst);
+            break;
```

Comment:
2 data copies are happened here.
We should avoid that to keep OpenCV effective library.

Use:
- `dst.create()`
- `data = dst.getMat()`
- fill `data`

First 2 calls could be done outside of switch for successful path.